### DISCUSSION TITLE
How to set up a vector index via sql

### DISCUSSION THREAD

**Discussion by danielcamara** [ArcadeData/arcadedb#1419] [Q&A]
Hi everyone, Happy New Year!

I have a query regarding vector information handling in ArcadeDB. While I've reviewed the documentation, I am afraid I need some additional details. I somewhat understand the Java code for creating an in-memory index for vectors, but as I'm not using Java, this example isn't directly applicable. The SQL example, as I understand it, assumes an existing index, as it is an  import statement (_import database https://dl.fbaipublicfiles.com/fasttext/vectors-crawl/cc.en.300.vec.gz with distanceFunction = 'cosine', m = 16, ef = 128, efConstruction = 128;_). 

Is there a method to create the index through the management interface or via SQL/cypher/gremlin..., ?  That would enable me to do that via the Python wrapper library!

Specifically, I need to know how to:

   - Create a database that take the vectors into account
   - Define different node types that will be affected by vectors (e.g., "Word" nodes in the example - _persistentIndex = hnswIndex.createPersistentIndex(database).withVertexType("Word").withEdgeType("Proximity").withVectorPropertyName("vector").withIdProperty("name").create();_).
   - Confirm that vectors are created based on the node's "name" property.
   - Ensure that every new "Word" node automatically includes the vector once the vector index is set.
   - Use Cypher for similarity searches, for example how to translate the following Java lines into Cypher:
     -- _persistentIndex = (HnswVectorIndex) database.getSchema().getIndexByName("Word[name,vector]");_
      -- _List<SearchResult<Vertex, Float>> approximateResults = persistentIndex.findNeighbors(input, k);_
   - Modify vector index properties, such as EfConstruction, on the fly after initial setup. I mean the value of EfConstruction is 200 in the beginning, but we dicide to change it to 300 for example, is it possible? Which would be the impact on the existing index?

Apologies if these questions seem basic or naive. If you have a comprehensive example encompassing the creation of a database, addition of an index, node insertion, and relationship assessment would be immensely helpful.

Best regards,

Daniel

**Comment by danielcamara** [comment]
I am sorry for bothering you again, but work with the vector without java is it possible, impossible.... if it is impossible it is fine to me! At this point I would just like to know :)

**Comment by lvca** [comment]
> Hi everyone, Happy New Year!

Happy New Year, sorry for the delay. I responded on Discord, but I haven't seen a discussion here.

> I have a query regarding vector information handling in ArcadeDB. While I've reviewed the documentation, I am afraid I need some additional details. I somewhat understand the Java code for creating an in-memory index for vectors, but as I'm not using Java, this example isn't directly applicable. The SQL example, as I understand it, assumes an existing index, as it is an import statement (_import database https://dl.fbaipublicfiles.com/fasttext/vectors-crawl/cc.en.300.vec.gz with distanceFunction = 'cosine', m = 16, ef = 128, efConstruction = 128;_).

The documentation about vectors is still poor, there are so many things to cover and missing examples. We're going to improve this asap.

For now, please check this source for more information:

https://github.com/ArcadeData/arcadedb/blob/1de993df1769606102964851e39bd314616e88ca/integration/src/test/java/com/arcadedb/integration/importer/vector/FastTextDatabase.java#L23-L22

The SQL command creates the index automatically if not existent.

> Is there a method to create the index through the management interface or via SQL/cypher/gremlin..., ? That would enable me to do that via the Python wrapper library!

Yes, SQL command.

> Specifically, I need to know how to:
> 
> * Create a database that take the vectors into account
> * Define different node types that will be affected by vectors (e.g., "Word" nodes in the example - _persistentIndex = hnswIndex.createPersistentIndex(database).withVertexType("Word").withEdgeType("Proximity").withVectorPropertyName("vector").withIdProperty("name").create();_).
> * Confirm that vectors are created based on the node's "name" property.
> * Ensure that every new "Word" node automatically includes the vector once the vector index is set.
> * Use Cypher for similarity searches, for example how to translate the following Java lines into Cypher:
>   -- _persistentIndex = (HnswVectorIndex) database.getSchema().getIndexByName("Word[name,vector]");_
>   -- _List<SearchResult<Vertex, Float>> approximateResults = persistentIndex.findNeighbors(input, k);_

Open Cypher doesn't support vectors in the syntax. You can use it via SQL and Java API only.

> * Modify vector index properties, such as EfConstruction, on the fly after initial setup. I mean the value of EfConstruction is 200 in the beginning, but we dicide to change it to 300 for example, is it possible? Which would be the impact on the existing index?

These values can't be changed after creation (any vector dbms has the same limitation). If you need to change it, then drop and recreate the index.

**Reply by danielcamara** [reply]
Hi I am sorry, I am not a Discord user... so I may have lost your answer. 
Thank you for the answer and the pointer.  I got the example, but it gives me an error. Apparently, the vector file should have an issue. It gives me a parsing error.
``
Exception in thread "main" com.arcadedb.exception.CommandSQLParsingException: import database file://C:\Users\Meryem\AppData\Local\Temp\cc.en.300.vec.gz with distanceFunction = cosine, m = 16, ef = 128, efConstruction = 128,vertexType = Word, edgeType = Proximity, vectorProperty = vector, vectorType = Float, idProperty = name
	at com.arcadedb.query.sql.parser.StatementCache.throwParsingException(StatementCache.java:130)
	at com.arcadedb.query.sql.parser.StatementCache.parse(StatementCache.java:118)
	at com.arcadedb.query.sql.parser.StatementCache.get(StatementCache.java:70)
	at com.arcadedb.query.sql.SQLQueryEngine.parse(SQLQueryEngine.java:208)
	at com.arcadedb.query.sql.SQLQueryEngine.command(SQLQueryEngine.java:106)
	at com.arcadedb.database.LocalDatabase.command(LocalDatabase.java:1319)
	at FastTextDatabase.<init>(FastTextDatabase.java:86)
	at FastTextDatabase.main(FastTextDatabase.java:56)
Caused by: com.arcadedb.query.sql.parser.TokenMgrError: Lexical error at line 1, column 26.  Encountered: "\\" (92), after : ""
	at com.arcadedb.query.sql.parser.SqlParserTokenManager.getNextToken(SqlParserTokenManager.java:5858)
	at com.arcadedb.query.sql.parser.SqlParser.jj_ntk_f(SqlParser.java:27456)
	at com.arcadedb.query.sql.parser.SqlParser.ImportDatabaseStatement(SqlParser.java:17152)
	at com.arcadedb.query.sql.parser.SqlParser.StatementInternal(SqlParser.java:1395)
	at com.arcadedb.query.sql.parser.SqlParser.Statement(SqlParser.java:1055)
	at com.arcadedb.query.sql.parser.SqlParser.Parse(SqlParser.java:251)
	at com.arcadedb.query.sql.parser.StatementCache.parse(StatementCache.java:111)
	... 6 more 
``

Even though the example is good, I understand now how to iterate over the database and how to search for close relations, once more, thank you.  
I am sorry if I am too slow, but as far as I understood, let's assume I managed to load the fasttext dataset. It is an image of a specific database, right? Or did I misunderstood it?  
That is really great, but now how do I create one from scratch for my own database, where the nodes have a different structure altogether?  

I am sorry for being insistent, but what syntax would be for the SQL to create the vector index, with my own data structure?
